### PR TITLE
Add normalisation factor to Integration for RebinnedOutput #33407

### LINKED ISSUE #33407: Integration with RebinnedOutput workspaces
**Original reporter:** R. Stewart / ISIS

**Describe the bug**

There is a maths error in the way in which the `Integration` algorithm handles `RebinnedOutput` workspaces which have fractional weights *Fi*. Taking a spectrum with *n* bins within the integration range, `Integration` is meant to sum up the **counts** within the range. For a `RebinnedOutput` the counts are defined by:

where *Yi* and *Fi* are the values stored in the `Y` and `F` vectors of the workspace. The current calculations compute:

However, if we take the limiting case where there are no bins with fractional weights and all the *Fi* values are unity, then we are actually computing:

So the computed integrals are actually smaller by a factor *n* (the number of bins being integrated over) than they should be. We should thus multiply (normalise) the integrated output by the number of bins.

However, this does not properly account for regions of the data where there is no overlap, where the *Fi* values are defined to be zero. In the `Integration` algorithm, these bins (which previously had Y values of `NaN`) have their values [set to zero](https://github.com/mantidproject/mantid/blob/715ccc45baa79904e33d433678ef1384b1c5792d/Framework/Algorithms/src/Integration.cpp#L382) and included in the sum, *but they should not be included in the normalisation* because they have by definition *Fi*=0. The normalisation factor is thus the number of non-zero-fraction bins rather than the number of bins.

**To Reproduce**

This maths error is most clearly seen in the following vanadium measurement, where there is only intensity in the elastic line, so logically if we integrate over the elastic line then the integral (as long as it covers the full elastic energy window) should be independent of the integration range.

The data file is here: [LET73024_3.70meV.nxspe.zip](https://github.com/mantidproject/mantid/files/7966616/LET73024_3.70meV.nxspe.zip), and this script will plot a 2D spectrum in Q-E (Momentum-Energy): [Note that the call to `SofQW3` produces a `RebinnedOutput` workspace with fractional bins].

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
import numpy as np

ws = Load('LET73024_3.70meV.nxspe')
ws_rebinnedoutput = SofQW3(ws, QAxisBinning='0,0.01,2', EAxisBinning='-3,0.02,3', EMode='Direct') 
fg, ax = plt.subplots(subplot_kw={'projection': 'mantid'})
ax.pcolormesh(ws_rebinnedoutput, norm=LogNorm() )
fg.show()
```
integration_bug_2d

As can be seen, outside of the range [-0.5, 0.5]meV the intensity is almost zero, so logically integrating from [-1,1]meV or [-2,2]meV should produce the same integrals. However, running this script after the previous:

```python
fg, ax = plt.subplots(subplot_kw={'projection': 'mantid'})
for ii, irange in enumerate([0.5, 1, 2]):
    ws_int = Integration(ws_rebinnedoutput, -irange, irange, OutputWorkspace=f'ws_int_{ii}')
    ws_int = Transpose(ws_int, OutputWorkspace=f'ws_int_{ii}')
    ax.errorbar(ws_int)
ax.set_ylim(0, 20)
fg.show()
```
integration_bug_1d

shows that not only does the different integration range give different values these values are also incorrect insofar as the data represents an absolute units measurement so that the integral should equal the cross-section of Vanadium - which is 406 mbarns / steradian. In the original call to `SofQW3` we used 0.02 meV bins so the [-0.5, 0.5] integration should have 50 bins and we see that multiplying the result by 50 gives approximately 450 mbarns/sr, which is close to the expected value (absolute intensity measurements are subject to systematic errors due to the absolute units corrections and are often only ~10% accurate).

In contrast, running the above loop with `Integration(ws, ...` instead of `Integration(ws_rebinnedoutput, ...` gives curves which lay on top of each other with values around 450 mbarns/sr as expected.

**Expected behavior**

The integrated values for a elastic line measurement should not depend on the integration range if the range is larger than the elastic window, and the integrated values of an absolute units measurements should be close to the expected cross-section of the materials measured.

**Screenshots**
N/A

**Platform/Version (please complete the following information):**
 - OS: All (tested on SL7 and Windows 10)
 - Mantid Version: Nightly (`6.2.20220127.1718`)

**Additional context**

### PR DESCRIPTION
**Pull request by mducle**
**Description of work.**

Adds a normalisation factor to the `Integration` calculation for `RebinnedOutput` to fix the maths error described in #33407. 

The issue is resolved if the counts are **multiplied** by the number of non-NaN bins integrated over, but for a `RebinnedOutput`, the `finalize`d counts is *Ci = Yi/Fi* - so in order to not change the code for the non-`RebinnedOutput` case or add another `if` block, we instead divide the total fractions `sumF` by the normalisation (since the intermediate Y values are divided by the summed fractions to give the final counts / distribution frequencies).

**To test:**

Run this script, derived from that in the issue (the data file is [here](https://github.com/mantidproject/mantid/files/7966616/LET73024_3.70meV.nxspe.zip)):

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
import numpy as np

ws = Load('LET73024_3.70meV.nxspe')
ws_rebinnedoutput = SofQW3(ws, QAxisBinning='0,0.01,2', EAxisBinning='-3,0.02,3', EMode='Direct') 
fg, ax = plt.subplots(subplot_kw={'projection': 'mantid'})
for ii, irange in enumerate([0.5, 1, 2]):
    ws_int = Integration(ws_rebinnedoutput, -irange, irange, OutputWorkspace='ws_int_%i' % (ii))
    ws_int = Transpose(ws_int, OutputWorkspace='ws_int_%i' % (ii))
    ax.errorbar(ws_int)
ax.set_ylim(0, 1000)
fg.show()
```
integration_bug_1d_fixed

And check that the three integrated curve are close to each other in value with values close to 400 mbarn/steradian

Fixes #33407. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/Integration.cpp [hunk 1]
```diff
@@ -270,6 +271,9 @@ void Integration::integrateSpectrum(const API::ISpectrum &inSpec, API::ISpectrum
     if (Fin) {
       // Workspace has fractional area information, need to take into account
       sumF = std::accumulate(Fin->begin() + distmin, Fin->begin() + distmax, 0.0);
+      // Need to normalise by the number of non-NaN bins - see issue #33407 for details
+      Fnor = static_cast<double>(
+          std::count_if(Fin->begin() + distmin, Fin->begin() + distmax, [](double i) { return i != 0.; }));
```
[the remark this answers is not in the document]
Comment (mducle) [reply Framework/Algorithms/src/Integration.cpp:276]: sure... what about `f`? Or something more verbose? (`i` was supposed to be for "input").

File: docs/source/release/v6.3.0/33408_framework.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+Algorithms
+----------
+
+Bug Fixes
+#########
+- Fixed a bug in :ref:`Integration <algm-Integration>` when using it with a ``RebinnedOutput`` workspace created by :ref:`SofQWNormalisedPolygon <algm-SofQWNormalisedPolygon>` where the output was not correctly normalised.
```
Comment (AnthonyLim23) [inline docs/source/release/v6.3.0/33408_framework.rst:6]:
```suggestion
- Fixed a bug in :ref:`Integration <algm-Integration>` when using it with a ``RebinnedOutput`` workspace (e.g from :ref:`SofQWNormalisedPolygon <algm-SofQWNormalisedPolygon>`) where the output was not correctly normalised.
```

**Comment by AnthonyLim23** [comment]
I think this is a nice change and it would be good to capture this in some of the documentation as it address a subtle point.

When I was using it, I have found that if I convert from distribution (see below script) all of the lines still lie on top of each other. However, the average value is about 5. Is this expected?

```
from mantid.simpleapi import *
import matplotlib.pyplot as plt
from matplotlib.colors import LogNorm
import numpy as np

ws = Load('LET73024_3.70meV.nxspe')
ws_rebinnedoutput = SofQW3(ws, QAxisBinning='0,0.01,2', EAxisBinning='-3,0.02,3', EMode='Direct') 
fg, ax = plt.subplots(subplot_kw={'projection': 'mantid'})
for ii, irange in enumerate([0.5, 1, 2,3,4,5,6]):
    ws_int = Integration(ws_rebinnedoutput, -irange, irange, OutputWorkspace='ws_int_%i' % (ii))
    ws_int = Transpose(ws_int, OutputWorkspace='ws_int_%i' % (ii))
    ConvertFromDistribution(ws_int)
    print(ws_int)
    ax.plot(ws_int.readX(0)[:-1], ws_int.readY(0), linestyle="--",label=f'-{irange} to {irange}')
#ax.set_ylim(0, 1000)
ax.legend()
fg.show()
```

File: Framework/Algorithms/src/Integration.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply Framework/Algorithms/src/Integration.cpp:276]: `f` would be better, I like verbose names personally. But for something internal like this it is less important

**Comment by mducle** [comment]
@AnthonyLim23 

> However, the average value is about 5

Hmmm... so when it is a distribution, the plotted values are "frequencies" and the units are strictly "per reciprocal Angstrom" - since we are using a step size of 0.01 A^-1, there is a factor of 100 between the "bin-width normalisation" when the workspace is a distribution compared to when it is point data. 

I'm not really sure what to do about this... on the one hand the distribution / point data distinction really only matters for ragged binning - for uniform bins like we have here it just means a constant scaling factor between the two - but this constant factor is important if we're talking about absolute units. 

Now, my personal feeling is that the conversion from distribution after running `Integration` is undefined behaviour. `Integration` is set up to give consistent behaviour if users stick to either distributions or point data - e.g. if they want point data it should be converted from distribution *before* calling `Integration`; conversely if they have used `Integration` on a distribution it should be kepted as a distribution afterwards... (if you run `ConvertFromDistribution` on `ws_rebinnedoutput` before the loop then `Integration` will give the "correct" point data (under the "none" normalisation)).

File: Framework/Algorithms/src/Integration.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mducle) [reply Framework/Algorithms/src/Integration.cpp:276]:
```suggestion
          std::count_if(Fin->begin() + distmin, Fin->begin() + distmax, [](double f) { return f != 0.; }));
```

**Comment by AnthonyLim23** [comment]
Just checked an doing the conversion before the integration gives the expected result

**Comment by mducle** [comment]
> it would be good to capture this in some of the documentation

I could add something to the `Integration` documentation. Alternatively I can write a new "Fractional Rebinning (RebinnedOutput workspaces)" concepts page, and change the docs for `SofQW3`, `Rebin2D`, `Integration` etc which can handle `RebinnedOutput` to reference the concept page...

**Comment by mducle** [comment]
It probably depends on the user - new users would benefit from a concepts page, more established users would probably not bother looking at that... Probably the most useful (and easiest to do for now) is to add to the `Integration` documentation now and make an issue to add a concepts page to the docs later...

**Review by AnthonyLim23 — APPROVED** [review]

**Pull request merged**